### ISSUE TITLE
Allow custom styling for consent banner

### ISSUE THREAD

**Issue by m90** [offen/offen#505]
Right now the consent banner is styled one-size-fits-all. While we try to keep it low-impact for most designs, this surely hurts adoption in cases where a unified UX is important for operators.

While it's technically trivial to allow operators to inject custom CSS into the iframe element that displays the banner, this also comes with some pitfalls and requirements:

- Operators should not be able to alter the content in the banner (i.e. change wording to be misleading or hide certain elements)
- The default a11y and privacy standards must be preserved in any case

## Approach: allowing operators to define custom CSS per account

Semantic classes are applied to the markup that displays the consent banner. Operators can now add custom CSS for each account which the `server` will add (in addition to the predefined styles) to each `vault` response. In order to prevent operators from creating a misleading consent banner the following sanitzation rules are applied to the CSS before saving:

- no pseudo content
- no hiding of elements
- no opacity rules

### Pros

- Allows operators to apply fine grained styling that matches virtually any site design
- Technically simple to implement

### Cons

- Sanitizing CSS will help, but it will not close all loopholes for malicious use ever (then again, forking Offen is a lot easier and gives you all options)
- Does not offer a middle ground for non-technically savvy users that don't want to / cannot write CSS but still want to customize their banner's appearace

**Comment by m90** [comment]
As discussed with @hendr-ik in person we would like to start off with the guidelines for sanitizing CSS: we trust the operator who deploys and customizes Offen to keep the consent guarantees intact but aim to prevent attackers (who might hijack a system or similar) from injecting malicious content through CSS.

This means we'll start with the following rules:

1. no loading of external resources (as this would introduce an external tracking vector)
2. No use of `content`. We want to prevent unwanted messages from showing up or attackers changing content.
3. No use of transparency as this would allow attackers to create an invisible banner and do "clickjacking".

**Comment by TommasoAmici** [comment]
Hi, I don't have the chance to give this much thought from a development point of view.

From a user point of view I really need simple customizations to make Ogden fit in with my designs: border, colors and maybe font family are all I need to touch

**Comment by m90** [comment]
> I don't have the time to give this much thought from a development point of view in this moment.

Your requirements as a user are super helpful and all we need. Thank you.

**Comment by whalehub** [comment]
@m90 I concur with @TommasoAmici. The consent banner currently sticks out like a sore thumb because its yellow color clashes with the dark design theme on my website. All it would take to fix that is the ability to change the color of the different banner elements.

**Comment by m90** [comment]
@TommasoAmici @whalehub We just merged the PR containing the feature and plan to include it in the next release.

In case you are interested and would like to help us ship a feature that is actually helpful, we'd appreciate any feedback on a test run. To check it out, you can run the `development` version in demo mode sth like this:

```console
docker pull offen/offen:development
docker run --rm -p 9876:9876  offen/offen:development demo -port 9876
```

After logging in head to `Customize appearance`, the documentation link is currently broken because this is not properly released yet, but it would point here https://docs.offen.dev/v/development/running-offen/customizing-consent-banner/

Our primary questions for feedback would be:

- can you achieve what you would like to do visually without too much hair pulling?
- would you need more/less/different documentation to get up and running?

Any other feedback is welcome too of course. Thank you!

**Comment by whalehub** [comment]
@m90 I love how you implemented this feature, what with the preview function and everything - well done! The documentation was more than sufficient for me personally.

**Before**

![img](https://xbb.datahoarder.dev/VUSu3/sOXOwIlo11.png/raw)

**After**

![img](https://xbb.datahoarder.dev/VUSu3/fOWOrERE32.png/raw)

**Comment by TommasoAmici** [comment]
Beautiful! The documentation is very good too

<img width="487" alt="image" src="https://user-images.githubusercontent.com/424525/116529951-c4ec6e80-a8dd-11eb-8e38-70497a46f433.png">

**Comment by m90** [comment]
This now released in [v0.4.2](https://github.com/offen/offen/releases/tag/v0.4.2). Thanks for all of your feedback :tophat:
